transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java

```diff
@@ -0,0 +1,897 @@
+
+import static org.ehcache.spi.ServiceLocator.findAmongst;
+import static org.ehcache.spi.ServiceLocator.findSingletonAmongst;
+
+/**
+ * A {@link Store} implementation wrapping another {@link Store} driven by a JTA
+ * {@link javax.transaction.TransactionManager} using the XA 2-phase commit protocol.
+ *
+ * @author Ludovic Orban
+ */
+public class XAStore<K, V> implements Store<K, V> {
+
+  private final Class<K> keyType;
+  private final Class<V> valueType;
+  private final Store<K, SoftLock<V>> underlyingStore;
+  private final TransactionManagerWrapper transactionManagerWrapper;
+  private final Map<Transaction, EhcacheXAResource<K, V>> xaResources = new ConcurrentHashMap<Transaction, EhcacheXAResource<K, V>>();
+  private final TimeSource timeSource;
+  private final Journal<K> journal;
+  private final String uniqueXAResourceId;
+  private final XATransactionContextFactory<K, V> transactionContextFactory;
+  private final EhcacheXAResource recoveryXaResource;
+
+  public XAStore(Class<K> keyType, Class<V> valueType, Store<K, SoftLock<V>> underlyingStore, TransactionManagerWrapper transactionManagerWrapper,
+                 TimeSource timeSource, Journal<K> journal, String uniqueXAResourceId) {
+    this.keyType = keyType;
+    this.valueType = valueType;
+    this.underlyingStore = underlyingStore;
+    this.transactionManagerWrapper = transactionManagerWrapper;
+    this.timeSource = timeSource;
+    this.journal = journal;
+    this.uniqueXAResourceId = uniqueXAResourceId;
+    this.transactionContextFactory = new XATransactionContextFactory<K, V>(timeSource);
+    this.recoveryXaResource = new EhcacheXAResource<K, V>(underlyingStore, journal, transactionContextFactory);
+  }
+
+  private boolean isInDoubt(SoftLock<V> softLock) {
+    return softLock.getTransactionId() != null;
+  }
+
+  private ValueHolder<SoftLock<V>> getSoftLockValueHolderFromUnderlyingStore(K key) throws CacheAccessException {
+    return underlyingStore.computeIfPresent(key, new BiFunction<K, SoftLock<V>, SoftLock<V>>() {
+      @Override
+      public SoftLock<V> apply(K k, SoftLock<V> vSoftLock) {
+        return vSoftLock;
+      }
+    });
+  }
+
+  private XATransactionContext<K, V> getCurrentContext() throws CacheAccessException {
+    try {
+      final Transaction transaction = transactionManagerWrapper.getTransactionManager().getTransaction();
+      if (transaction == null) {
+        throw new XACacheAccessException(new XACacheException("Cannot access XA cache outside of XA transaction scope"));
+      }
+      EhcacheXAResource<K, V> xaResource = xaResources.get(transaction);
+      if (xaResource == null) {
+        xaResource = new EhcacheXAResource<K, V>(underlyingStore, journal, transactionContextFactory);
+        transactionManagerWrapper.registerXAResource(uniqueXAResourceId, xaResource);
+        transactionManagerWrapper.getTransactionManager().getTransaction().enlistResource(xaResource);
+        xaResources.put(transaction, xaResource);
```

Comment:
A transaction can only ever be bound/visible by one thread at a time, though it can migrate from one thread to another by the means of explicit calls to TransactionManager.suspend() / .resume(). And we create a new EhcacheXAResource instance for every transaction.